sql: fix system database survival on create or drop database

jeffswenson opened the pull request:
Run the system database region survival promotion/demotion logic when creating or dropping databases. This fixes a bug where creating a database that survives region failures would not promote the system database to survive region.

This logic is temporary and should be removed once RFC #109843 allows us to configure the system database as a mix of survive zone and survive region.

Fixes: #109843

jeffswenson committed:
sql: fix system database survival on create or drop database

Run the system database region survival promotion/demotion logic when
creating or dropping databases. This fixes a bug where creating a
database that survives region failures would not promote the system
database to survive region.

This logic is temporary and should be removed once RFC #109843 allows us
to configure the system database as a mix of survive zone and survive
region.

Fixes: #109843

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/109844)
<!-- Reviewable:end -->

JeffSwenson commented:
Shout out to @chrisseto for helping me find the missing `db.Dropped()` call in `maybeUpdateSystemDBSurvivalGoal`.

fqazi reviewed (approved):
<img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>...minus the CI issues

> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/109844#-:-NdahSFxDpNfAZqbpT5b:bjplv5k)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @JeffSwenson)



<!-- Sent from Reviewable.io -->

fqazi reviewed (approved):
> Reviewed 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/109844)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @JeffSwenson)



<!-- Sent from Reviewable.io -->

JeffSwenson commented:
I fixed a few declarative schema changer tests. The fallback in the declarative schema changer's drop database implementation required reading the system database descriptor. A few of the schema tests inject a custom set of descriptors which did not include the system database. I fixed the test by adding the system database to the test's descriptor catalog.

JeffSwenson commented:
bors r+
